### PR TITLE
[AMD] [ROCm] [Optimum] Add optimum-amd support

### PR DESCRIPTION
**Pull request by tjtanaa**
# Description
Add `optimum-amd` support to `infinity_emb`.

**Note:** To use `optimum-amd`, it is recommended to 
1. build from docker image `docker build -f libs/infinity_emb/Dockerfile.amd_auto -t ghcr.io/embeddedllm/infinity-rocm:opti
mum-amd ./libs/infinity_emb`
2. Pull a docker image `docker pull ghcr.io/embeddedllm/infinity-rocm:optimum-amd` (This is a docker image of this PR. Hopefully in the newer version of infinity, the optimum-amd support can be found in https://hub.docker.com/r/michaelf34/infinity ) So for now, as a quickstart, get it from EmbeddedLLM.

To launch the docker container:
* Interactive mode:
  * Launch docker container
    ```bash
    #!/bin/bash
    
    docker run -it \
      --cap-add=SYS_PTRACE \
      --security-opt seccomp=unconfined \
      --device=/dev/kfd \
      --device=/dev/dri/renderD128 \
      --device=/dev/dri/renderD136 \
      --group-add video \
      --network host \
      --entrypoint /bin/bash \
      ghcr.io/embeddedllm/infinity-rocm:optimum-amd \
      -c "source .venv/bin/activate && bash"
      ```
   * Launch Embedding Model
      ```bash
      HIP_VISIBLE_DEVICES=0 infinity_emb v2 --port 6909 --model-id BAAI/bge-m3 --model-warmup --device cuda  --engine optimum 
      ```
* Single line:
```bash
  #!/bin/bash
  
  docker run -it \
    --cap-add=SYS_PTRACE \
    --security-opt seccomp=unconfined \
    --device=/dev/kfd \
    --device=/dev/dri/renderD128 \
    --device=/dev/dri/renderD136 \
    --group-add video \
    --network host \
    --entrypoint /bin/bash \
    ghcr.io/embeddedllm/infinity-rocm:optimum-amd \
    -c "(source .venv/bin/activate) && (HIP_VISIBLE_DEVICES=0 infinity_emb v2 --port 6909 --model-id BAAI/bge-m3 --model-warmup --device cuda  --engine optimum)"
```

# CHANGES
* `libs/infinity_emb/Dockerfile.amd_auto`
  * Added installation step of `onnxruntime-rocm` and `optimum-amd`.

# Performance
These are the steps obtained in the warm-up run of the infinity server

| Configuration |  Performance |
|-----------------|----------------|
| Torch only | `model warmed up, between 476.69-2408.85 embeddings/sec at batch_size=32` |
| Torch Compile | `model warmed up, between 487.85-2789.83 embeddings/sec at batch_size=32` |
| Optimum AMD | `model warmed up, between 268.33-4903.16 embeddings/sec at batch_size=32` |

Running `benchmark_embed`
| Model                                       | Requests # / sec (mean) | Time (seconds) |
|---------------------------------------------|-------------------------|----------------|
| infinity (torch + no compile + fa2 disabled)| 2.52                    | 3.965         |
| infinity (torch +  compile + fa2 disabled)  | 0.52 (first run), 2.84 (second run)    | 18.612   , 3.517      |
| infinity (optimum-amd)  |  1.33    | 7.523      |

## Torch Only
torch-only

## Torch Compile
torch-compile

## Optimum-AMD
Screenshot 2024-10-26 220350
Screenshot 2024-10-26 220308
Screenshot 2024-10-26 220442

### CODE REVIEW

**Comment by tjtanaa** [comment]
@michaelfeil 

| Model                                       | Requests # / sec (mean) | Time (seconds) |
|---------------------------------------------|-------------------------|----------------|
| infinity (torch + no compile + fa2 disabled)| 2.52                    | 3.965         |
| infinity (torch +  compile + fa2 disabled)  | 0.52 (first run), 2.84 (second run)    | 18.612   , 3.517      |
| infinity (optimum-amd)  |  1.33    | 7.523      |

Will running the same data samples over the infinity embedding server introduced biases to the benchmark value of torch compile model? I have launched the embedding server with warm-up, yet why there is a huge difference between two benchmark runs?

**Comment by michaelfeil** [comment]
@tjtanaa I ignore the bots comments for style, but every 1 in 10 comments is useful. 


I added some options to improve, e.g. the dockerfile. If you dont have the capacity to work on it, I can do this changes in a couple of days. Thanks for the contribution again.

On which hardware did you run the above benchmarks? FYI, the --no-bettertransformer flag just disables `torch.nested flash-attention` which is not supported on amd.

**Comment by tjtanaa** [comment]
> @tjtanaa I ignore the bots comments for style, but every 1 in 10 comments is useful.
> 
> I added some options to improve, e.g. the dockerfile. If you dont have the capacity to work on it, I can do this changes in a couple of days. Thanks for the contribution again.
> 
> On which hardware did you run the above benchmarks? FYI, the --no-bettertransformer flag just disables `torch.nested flash-attention` which is not supported on amd. Amd should be still using a decent version of sdpa https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html

The benchmark is ran on MI300X.

**Comment by michaelfeil** [comment]
@tjtanaa How to continue from here? My Laptop breaks down on building the wheel from scratch and the rocm  wheel that is pre-built is for Radeon and not for MI Series.
